server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -305,6 +352,9 @@ public static class Balancer {
         private final double avgWriteLoadPerNode;
         private final double avgDiskUsageInBytesPerNode;
         private final NodeSorter sorter;
+        private final Map<DiscoveryNodeRole, Float> avgShardsPerNodeInTier;
+        private final Map<DiscoveryNodeRole, Double> avgWriteLoadPerNodeInTier;
+        private final Map<DiscoveryNodeRole, Double> avgDiskUsagePerNodeInTier;
```

Comment:
NIT: I think we could store those stats grouped (the same way they are computed): `Map<DiscoveryNodeRole, TierAllocationStats> statsPerTier`.

Last time I tried to benchmark allocator it showed that most of the time spent is inside of weight function computing hash codes (for getting nodes and indices). I think it would be better if we could get all stats in on hash get call to minimize additional overhead.